### PR TITLE
Fix CP Partial Least Square

### PR DESCRIPTION
**Pull request by cyrillustan**
This fixes some implementation issues in `CP_PLSR()`. I also added the calculation of $R^2$ of X and Y, very useful parameters to inspect the performance of the model.

### CODE REVIEW

File: tensorly/regression/cp_plsr.py [hunk 1]
```diff
@@ -97,6 +99,7 @@ def fit(self, X, Y):
         self.Y_factors = [
             T.zeros((l, self.n_components), **T.context(X)) for l in T.shape(Y)
         ]
+        self.coef_ = T.zeros((self.n_components, self.n_components), **T.context(X))
```
Comment (aarmey) [inline tensorly/regression/cp_plsr.py:102]: What is `coef_`? Can you provide a comment?

File: tensorly/regression/cp_plsr.py [hunk 2]
```diff
@@ -178,14 +188,29 @@ def predict(self, X):
...
-        factors_kr = khatri_rao(self.X_factors, skip_matrix=0)
-        unfolded = unfold(X, 0)
-        scores = T.lstsq(factors_kr, T.transpose(unfolded))[0]  # = Tnew
-        estimators = T.lstsq(self.X_factors[0], self.Y_factors[0])[0]
-        return (
-            T.dot(
-                T.dot(T.transpose(scores), estimators), T.transpose(self.Y_factors[1])
+        X_projection = T.zeros((T.shape(X)[0], self.n_components), **T.context(X))
+        for component in range(self.n_components):
+            X_projection = T.index_update(
```
Comment (aarmey) [inline tensorly/regression/cp_plsr.py:193]: Does this mean that the Bro paper algorithm did in fact work in the end? I'm not seeing you subtracting from X_orig like we had previously discussed...

File: tensorly/regression/tests/test_cp_plsr.py [hunk 3]
```diff
@@ -207,3 +213,49 @@ def test_optimized_covariance(n_latent):
         )
 
     assert_allclose(max_cov, pls_cov)
+
+
+@pytest.mark.parametrize("n_response", [5, 7, 9])
```
Comment (aarmey) [inline tensorly/regression/tests/test_cp_plsr.py:218]: I don't see you using `n_response`. Can this parameterization be removed?

File: tensorly/regression/cp_plsr.py [hunk 4]
```diff
@@ -34,7 +34,7 @@ class CP_PLSR:
     """
 
     def __init__(
-        self, n_components, tol=1.0e-9, n_iter_max=100, random_state=None, verbose=False
+        self, n_components, tol=1.0e-8, n_iter_max=100, random_state=None, verbose=False
```
Comment (aarmey) [inline tensorly/regression/cp_plsr.py:37]: The interaction within a component was going on up to `n_iter_max` with the old tolerance and the tensorflow backend. I'm not sure why the accuracy is different in a backend-specific way, but this was helping with testing taking a more reasonable amount of time.

**Review by aarmey — CHANGES_REQUESTED** [review]
@cyrillustan thanks for fixing this. Here are a few questions to start review.

File: tensorly/regression/cp_plsr.py (hunk 1 above)
Comment (cyrillustan) [reply tensorly/regression/cp_plsr.py:102]: These are the coefficients of the linear model, the same as $\beta$ in ordinary least square $Y = X \beta + \epsilon$. This was notated in Bro's paper as $\mathbf{b}$. This parameter has always existed in Bro's paper, but was previously dismissed by me as trivial since there was no explanation for it.

File: tensorly/regression/cp_plsr.py (hunk 2 above)
Comment (cyrillustan) [reply tensorly/regression/cp_plsr.py:193]: Yes. The subtracting from the original Y turned out to be not a problem at all given the 0 columns elsewhere. Every component explains exactly what we have subtracted each round, in this case, the project of X to Y. Therefore, the definition of $R^2$ is rightfully different for X and Y; they are not symmetrical (surprise, surprise).

File: tensorly/regression/tests/test_cp_plsr.py (hunk 3 above)
Comment (cyrillustan) [reply tensorly/regression/tests/test_cp_plsr.py:218]: I have added it in the function so it will make sense now.

**Comment by aarmey** [comment]
Thanks @cyrillustan. I also made an adjustment to the CP factorization so that it is faster. This is a pretty big change, so I will wait for one other person to review before merging.

File: tensorly/regression/cp_plsr.py [hunk 5]
```diff
@@ -86,6 +98,8 @@ def fit(self, X, Y):
         # Mean center the data, record info the object
         self.X_shape_ = T.shape(X)
         self.Y_shape_ = T.shape(Y)
+        self.original_X_ = T.copy(X)
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:101]: I'm not sure I'd want to store the full X and Y, it seems unnecessary and potentially very expensive (memory-wise): i'd let the user take care of that

File: tensorly/regression/cp_plsr.py [hunk 6]
```diff
@@ -151,17 +174,24 @@ def fit(self, X, Y):
                 self.Y_factors[1], T.index[:, component], comp_Y_factors_1
             )
 
+            B = T.lstsq(self.X_factors[0], T.reshape(comp_Y_factors_0, (-1, 1)))[0]
+            self.coef_ = T.index_update(
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:178]: As I'm reviewing this, given your other PR @aarmey I guess if we deprecate tensorflow we'll be able to remove those pesky index_updates (I think JAX supports the syntactic sugar now)

File: tensorly/regression/cp_plsr.py [hunk 7]
```diff
@@ -178,14 +208,29 @@ def predict(self, X):
...
+            X_projection = T.index_update(
+                X_projection,
+                T.index[:, component],
+                multi_mode_dot(
+                    X,
+                    [factor[:, component] for factor in self.X_factors[1:]],
+                    range(1, T.ndim(X)),
+                ),
             )
+            X -= CPTensor(
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:222]: Creating a CP tensor for this will add unnecessary overhead: we can directly use [`tl.tenalg.outer`](https://tensorly.org/dev/modules/generated/tensorly.tenalg.outer.html#tensorly.tenalg.outer) instead.

File: tensorly/regression/cp_plsr.py [hunk 8]
```diff
@@ -256,11 +301,11 @@ def transform(self, X, Y=None):
 
                 Y -= T.dot(
                     T.dot(
-                        T.lstsq(T.transpose(X_scores), T.transpose(X_scores))[0],
-                        Y_scores[:, [component]],
+                        X_scores,
+                        T.reshape(self.coef_[:, component], (-1, 1)),
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:305]: Do we need the reshape here?

File: tensorly/regression/cp_plsr.py [hunk 9]
```diff
@@ -256,11 +301,11 @@ def transform(self, X, Y=None):
...
                     ),
-                    T.transpose(self.Y_factors[1][:, [component]]),
-                )  # Y -= T pinv(T) u q'
+                    T.transpose(T.reshape(self.Y_factors[1][:, component], (-1, 1))),
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:307]: Same, do we need the reshape? It's just a vector.

File: tensorly/regression/cp_plsr.py [hunk 10]
```diff
@@ -283,3 +328,33 @@ def fit_transform(self, X, Y):
             Return `x_scores` if `Y` is not given, `(x_scores, y_scores)` otherwise.
         """
         return self.fit(X, Y).transform(X, Y)
+
+    def X_r2_score(self):
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:332]: I would explicitly pass X here, and not store it, see first comment.

File: tensorly/regression/cp_plsr.py [hunk 11]
```diff
@@ -283,3 +328,33 @@ def fit_transform(self, X, Y):
...
+            the amount of variance in X explained by the model
+
+        """
+        X_original = self.original_X_ - self.X_mean_
+        X_reconstructed = CPTensor([None, self.X_factors]).to_tensor()
+        return (
+            1 - T.norm(X_reconstructed - X_original) ** 2.0 / T.norm(X_original) ** 2.0
+        )
+
+    def Y_r2_score(self):
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:347]: Similarly, I think we can pass Y explicitly.

**Review by JeanKossaifi — COMMENTED** [review]
Thanks, this looks good. 

I left some comments. I think we can move the R2 to the metrics subpackage (it's the same for both Y and X and can be useful for other classes too), and just compute it internally and store the results rather than the input X and Y. 

For the API, either:
1. we pass both X, X_mean and X_pred
2. we directly pass in an (optionally) centered X
3. we make X_mean optional and with a default of None

What do you think? I'd suggest following scikit-learn for this.

File: tensorly/regression/cp_plsr.py (hunk 5 above)
Comment (cyrillustan) [reply tensorly/regression/cp_plsr.py:101]: We can just calculate these numbers and save them in the object, but not save the `original_X` (only use them on the run). This is similar to how sklearn does.

File: tensorly/regression/cp_plsr.py (hunk 8 above)
Comment (cyrillustan) [reply tensorly/regression/cp_plsr.py:305]: `T.dot` will somehow report error on shape if it is not in a matrix.

File: tensorly/regression/cp_plsr.py (hunk 9 above)
Comment (cyrillustan) [reply tensorly/regression/cp_plsr.py:307]: Same, but `T.transpose(T.reshape)` is redundant indeed.

File: tensorly/regression/cp_plsr.py (hunk 7 above)
Comment (cyrillustan) [reply tensorly/regression/cp_plsr.py:222]: Okay, I'll change it to `cp_to_tensor()`

**Comment by cyrillustan** [comment]
I separate the R2 function out of this class to `metric`. Hopefully it is in the appropriate file. I just removed the functions of R2 X and Y entirely and make them attributes, as it makes more sense to calculate them along the fitting process, and we don't have to save or pass the original tensor. This resembles how sklearn does variance explained, too.

File: tensorly/regression/cp_plsr.py [hunk 12]
```diff
@@ -86,6 +103,8 @@ def fit(self, X, Y):
         # Mean center the data, record info the object
         self.X_shape_ = T.shape(X)
         self.Y_shape_ = T.shape(Y)
+        original_X_ = T.copy(X)
+        original_Y_ = T.copy(Y)
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:107]: Can we get around the creation of those copy? Can we just work always on the centered versions of X and Y (we can uncenter if needed as we have the mean)?

File: tensorly/regression/cp_plsr.py (hunk 1 above)
Comment (JeanKossaifi) [reply tensorly/regression/cp_plsr.py:123]: This is quite useful -- could we add it to the docstrings (e.g. in Notes) or the doc?

File: tensorly/regression/cp_plsr.py [hunk 13]
```diff
@@ -151,17 +181,40 @@ def fit(self, X, Y):
...
+                self.coef_,
+                T.index[:, component],
+                T.reshape(B, (-1,)),
+            )
+
             # Deflation
-            X -= CPTensor(
-                (None, [T.reshape(ff, (-1, 1)) for ff in comp_X_factors])
-            ).to_tensor()
+            X -= cp_to_tensor((None, [T.reshape(ff, (-1, 1)) for ff in comp_X_factors]))
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:192]: Why not use tl.tenalg.outer? It's just a rank-1 CP, i.e. a single outer product of vectors. That may be faster as avoiding the checks and creation of None weights to handle.

File: tensorly/regression/cp_plsr.py [hunk 14]
```diff
@@ -177,15 +230,30 @@ def predict(self, X):
...
+                    range(1, T.ndim(X)),
+                ),
             )
+            X -= cp_to_tensor((
+                    None,
+                    [T.reshape(X_projection[:, component], (-1, 1))]
+                    + [
+                        T.reshape(factor[:, component], (-1, 1))
+                        for factor in self.X_factors[1:]
+                    ],
```
Comment (JeanKossaifi) [inline tensorly/regression/cp_plsr.py:252]: Same here, we could just do the tl.tenalg.outer if it proves faster? Might be more readable too?

**Review by JeanKossaifi — COMMENTED** [review]
Thanks @cyrillustan! The R2 is in the right place indeed and you even added a unit-test, perfect!

Looks good to me, just a minor thing about using outer-product rather than cp_to_tensor, which might be a little faster/more readable as we're only substracting a rank-1 tensor, i.e. a simple outer-product of vectors.

File: tensorly/metrics/tests/test_regression.py [hunk 15]
```diff
@@ -21,6 +21,14 @@ def test_RMSE():
     assert_array_almost_equal(RMSE(y_true, y_pred), true_mse)
 
 
+def test_R2_score():
+    """Test for RMSE"""
+    X_original = np.random.rand(5, 4, 3)
```
Comment (JeanKossaifi) [inline tensorly/metrics/tests/test_regression.py:26]:
```suggestion
    X_original = tl.rand(5, 4, 3)
```

File: tensorly/regression/cp_plsr.py (hunk 12 above)
Comment (cyrillustan) [reply tensorly/regression/cp_plsr.py]: Hmm, I'm not sure how to avoid making a copy, either the original or the mean-centered version (what I changed into) of X and Y, if we still want to keep the $R^2$ calculation though. @aarmey do you have any idea?

Comment (aarmey) [reply tensorly/regression/cp_plsr.py]: I think Jean's point is that right now you have two persistent copies—of both the original and mean-centered. Just keep a copy of one, and then build the other if/where you need it.

Comment (cyrillustan) [reply tensorly/regression/cp_plsr.py]: Oh, then the latest change should make everything only mean-centered.

File: tensorly/metrics/tests/test_regression.py [hunk 16]
```diff
@@ -21,6 +21,14 @@ def test_RMSE():
     assert_array_almost_equal(RMSE(y_true, y_pred), true_mse)
 
 
+def test_R2_score():
+    """Test for RMSE"""
+    X_original = tl.randn((5, 4, 3))
+    assert R2_score(X_original, X_original) == 1.0
+    assert R2_score(X_original, X_original * 2) == 0.0
+    assert R2_score(X_original, np.zeros_like(X_original)) == 0.0
```
Comment (JeanKossaifi) [inline tensorly/metrics/tests/test_regression.py:29]:
```suggestion
    assert R2_score(X_original, tl.zeros_like(X_original)) == 0.0
```

**Comment by JeanKossaifi** [comment]
Thinking about this more, how about calling the R2 `r2_score` for consistency with sklearn, and, as they do, just extracting its computation from the regression class itself @aarmey @cyrillustan?

We can then either 
* let the users compute the r2 score themselves `r2_score(ground-truth, prediction)` 
* and/or add a `score` method that takes as input both the prediction and ground-truth and essentially just acts as an alias to r2 (probably can handle centering etc as well)

**Comment by cyrillustan** [comment]
I think my edits should fix the problem in `score()`. Let me know if there is anything else I should do. I see that some tests failed, but they don't seem to be related to `CP_PLSR` edits.

File: tensorly/regression/cp_plsr.py [hunk 17]
```diff
@@ -283,3 +317,18 @@ def fit_transform(self, X, Y):
...
+
+        Parameters
+        ----------
+        X : ndarray
+            tensor data of shape (n_samples, N1, ..., NS), same dimension as the X
+            in self.fit() all except the first dimension
+        Y : 2D-array of shape (n_samples, n_predictions)
+            the ground truth labels associated with each sample
+        """
+        from ..metrics.regression import R2_score
```
Comment (aarmey) [inline tensorly/regression/cp_plsr.py:332]: Is there a reason this is imported just at the function level?

**Review by aarmey — CHANGES_REQUESTED** [review]
One more question, then I am comfortable merging this.

**Comment by XiaoXiao-Woo** [comment]
作业已收到，祝同学生活快乐

File: tensorly/regression/cp_plsr.py (hunk 17 above)
Comment (cyrillustan) [reply tensorly/regression/cp_plsr.py:332]: Oh this is just mimicking how [sklearn did it](https://github.com/scikit-learn/scikit-learn/blob/9aaed498795f68e5956ea762fef9c440ca9eb239/sklearn/base.py#L720).
